src/Microsoft.AspNetCore.Hosting/GenericHost/GenericWebHostedService.cs

```diff
@@ -0,0 +1,200 @@
+
+        public GenericWebHostService(IOptions<GenericWebHostServiceOptions> options,
+                                     IServer server,
+                                     ILogger<GenericWebHostService> logger,
+                                     DiagnosticListener diagnosticListener,
+                                     IHttpContextFactory httpContextFactory,
+                                     IApplicationBuilderFactory applicationBuilderFactory,
+                                     IEnumerable<IStartupFilter> startupFilters,
+                                     IConfiguration configuration,
+                                     IHostingEnvironment hostingEnvironment)
+        {
+            Options = options.Value;
+            Server = server;
+            Logger = logger;
+            DiagnosticListener = diagnosticListener;
+            HttpContextFactory = httpContextFactory;
+            ApplicationBuilderFactory = applicationBuilderFactory;
+            StartupFilters = startupFilters;
+            Configuration = configuration;
+            HostingEnvironment = hostingEnvironment;
+        }
+
+        public GenericWebHostServiceOptions Options { get; }
+        public IServer Server { get; }
+        public ILogger<GenericWebHostService> Logger { get; }
+        public DiagnosticListener DiagnosticListener { get; }
+        public IHttpContextFactory HttpContextFactory { get; }
+        public IApplicationBuilderFactory ApplicationBuilderFactory { get; }
+        public IEnumerable<IStartupFilter> StartupFilters { get; }
+        public IConfiguration Configuration { get; }
+        public IHostingEnvironment HostingEnvironment { get; }
+
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            HostingEventSource.Log.HostStart();
+
+            var serverAddressesFeature = Server.Features?.Get<IServerAddressesFeature>();
+            var addresses = serverAddressesFeature?.Addresses;
+            if (addresses != null && !addresses.IsReadOnly && addresses.Count == 0)
+            {
+                var urls = Configuration[WebHostDefaults.ServerUrlsKey] ?? Configuration[DeprecatedServerUrlsKey];
+                if (!string.IsNullOrEmpty(urls))
+                {
+                    serverAddressesFeature.PreferHostingUrls = WebHostUtilities.ParseBool(Configuration, WebHostDefaults.PreferHostingUrlsKey);
+
+                    foreach (var value in urls.Split(new[] { ';' }, StringSplitOptions.RemoveEmptyEntries))
+                    {
+                        addresses.Add(value);
+                    }
+                }
+            }
+
+            RequestDelegate application = null;
+
+            try
+            {
+                var builder = ApplicationBuilderFactory.CreateBuilder(Server.Features);
+                Action<IApplicationBuilder> configure = Options.ConfigureApplication ?? _defaultApplication;
```

Comment:
Nevermind, I misread the signature of _defaultApplication, I thought it was throwing a per-request exception.  It's a little strange you wouldn't just throw the exception here rather than running the filters first. Do we have an examples of terminal filters that would would replace the app, or error handling filters that would catch the exception?

Reply:
This scenario wasn't supported by the old WebHost. If you didn't specify startup then the filters never ran. https://github.com/aspnet/Hosting/blob/d7b9fd480765bdc01f06441f308fb288e6001049/src/Microsoft.AspNetCore.Hosting/Internal/WebHost.cs#L191-L196

Reply:
So this behavior is important enough that a test should fail right?